### ISSUE TITLE
Synchronizer Stream Restarting on Error

### ISSUE THREAD

**Issue by Marpyre** [LuatixHQ/opencti#2623]
## Description

I have 2 instances of OpenCTI. One has all the connectors (Instance 1), the other (Instance 2) is ingesting the data from Instance 1 using a synchronizer. On Instance 1 (sender), the stream processor and Redis client start and close every minute. Approximately every hour, the entire container crashes with a JavaScript heap out of memory error. When the synchronizer is turned off on Instance 2, the errors on Instance 1 stop. This is related to ticket #2432.

## Environment

1. Ubuntu 16.4 on Azure
2. 5.3.17
3. OpenCTI client: { e.g. frontend or python }
4. Other environment details:

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. On Instance 1 (Sender) create Role- Sync with only Connect and consume the platform streams (/stream, /stream/live) permission. Create Group with Permissions to All Markings. Create User and add them to Sync role and Group.
2. On Instance 2 (Receiver) Add synchronizer using URL for Instance 1 and credentials for Sync user that was created.
3. Start synchronizer

## Expected Output

Synchronizer ingests all the data from Instance 1 to Instance 2.

## Actual Output

Instance 2 stops receiving data and Instance 1 crashes and restarts continually.

## Additional information

Logs from Instance 1 when streams are restarting every minute:

{"category":"APP","level":"info","message":"[STREAM] Listening stream live from Fri Nov 25 2022 15:08:11 GMT+0000","timestamp":"2022-12-01T15:34:39.690Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Starting stream processor for synch_user@email.com","timestamp":"2022-12-01T15:34:39.690Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[REDIS] Redis 'synch_user@email.com' client ready","timestamp":"2022-12-01T15:34:39.693Z","version":"5.3.17"}
{"auth":{...},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200779-brazilian-judicial-investigations-ramp-up-with-federal-police-issuing-21-search-and-seizure-warrants.pdf","metadata":{"entity_id":"f48e4dac-a834-4538-ab3b-55db31594a61","version":"2022-12-01T15:34:42.381Z"},"path":"import/Report/f48e4dac-a834-4538-ab3b-55db31594a61"},"timestamp":"2022-12-01T15:34:42.381Z","version":"5.3.17"}
{"auth":{...},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200781-ocean-buffalo-identified-using-highly-targeted-strategic-web-compromise-tactics.pdf","metadata":{"entity_id":"1db6e797-c0b5-4e6a-a8ac-b23f259fb3de","version":"2022-12-01T15:34:54.001Z"},"path":"import/Report/1db6e797-c0b5-4e6a-a8ac-b23f259fb3de"},"timestamp":"2022-12-01T15:34:54.001Z","version":"5.3.17"}
{"auth":{...},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200782-publicly-reported-operation-interception-demonstrates-ttp-overlaps-with-labyrinth-chollima.pdf","metadata":{"entity_id":"9dae027e-d466-4f8f-abc9-87dd249d2f26","version":"2022-12-01T15:35:06.942Z"},"path":"import/Report/9dae027e-d466-4f8f-abc9-87dd249d2f26"},"timestamp":"2022-12-01T15:35:06.943Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Closing stream processor for 8e5d9049-2777-471c-8df0-b3abc9183241","timestamp":"2022-12-01T15:35:38.649Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Shutdown stream processor for synch_user@email.com","timestamp":"2022-12-01T15:35:38.650Z","version":"5.3.17"}
{"auth":{"email":"synch_user@email.com","ip":"::ffff:10.42.0.211","user_id":"28ee9247-9935-4a2b-ae8f-9392fb90091c"},"category":"AUDIT","level":"info","message":"LOGIN","resource":{"provider":"Bearer"},"timestamp":"2022-12-01T15:35:51.600Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Listening stream live from Fri Nov 25 2022 15:08:11 GMT+0000","timestamp":"2022-12-01T15:35:53.091Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Starting stream processor for synch_user@email.com","timestamp":"2022-12-01T15:35:53.092Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[REDIS] Redis 'synch_user@email.com' client ready","timestamp":"2022-12-01T15:35:53.094Z","version":"5.3.17"}

Logs from Instance 1 when the container crashes:

{"auth":{"email":"synch_user@email.com","ip":"::ffff:10.42.0.211","user_id":"28ee9247-9935-4a2b-ae8f-9392fb90091c"},"category":"AUDIT","level":"info","message":"LOGIN","resource":{"provider":"Bearer"},"timestamp":"2022-12-01T16:21:12.277Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Listening stream live from Fri Nov 25 2022 15:08:11 GMT+0000","timestamp":"2022-12-01T16:21:18.032Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[STREAM] Starting stream processor for synch_user@email.com","timestamp":"2022-12-01T16:21:18.032Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[REDIS] Redis 'synch_user@email.com' client ready","timestamp":"2022-12-01T16:21:18.036Z","version":"5.3.17"}
<--- Last few GCs --->
[6:0x7fbb32a4e400] 4629696 ms: Scavenge (reduce) 8031.9 (8142.5) -> 8031.4 (8142.8) MB, 3.9 / 0.0 ms (average mu = 0.163, current mu = 0.027) allocation failure
[6:0x7fbb32a4e400] 4629703 ms: Scavenge (reduce) 8032.2 (8142.8) -> 8031.7 (8143.0) MB, 3.8 / 0.0 ms (average mu = 0.163, current mu = 0.027) allocation failure
[6:0x7fbb32a4e400] 4629710 ms: Scavenge (reduce) 8032.4 (8143.0) -> 8032.0 (8143.3) MB, 4.3 / 0.0 ms (average mu = 0.163, current mu = 0.027) allocation failure
<--- JS stacktrace --->
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

**Comment by richard-julien** [comment]
Hi @Marpyre , difficult to answer looks like your platform 1 is overloaded. Depending of the amount of connectors and workers you have + stream sync, its possible that your platform is undersized. When i say undersized you have multiple approach to scale depending of your archiecture. You can obsvisouly give more memory to the node process, but you can also deploy multiple api to split the load between ingestion, users and streaming.

**Comment by Marpyre** [comment]
Hi @richard-julien, we should have plenty of memory given to the node process, and the vm never exceeds 50% memory utilization. I think the out of memory issue is a tangential symptom of a core synchronizer issue. Also, is there any documentation on the "deploy multiple api to split the load" that I could reference in case that might improve our setup?

**Comment by Marpyre** [comment]
I realized by setting NODE_OPTIONS=--max-old-space-size=8096, there was a max memory that would cause the process to crash before using the rest of the memory on the vm. I will definitely increase this limit in case that is the issue. Eventually got past the error with enough time and several container restarts but there is a bunch of missing data. I tried running the synchronizer back over the time period with missing data and I have additional comments/questions:

On **Instance 1** (sender OpenCTI instance): I see this in the logs relatively frequently:

```
(node:7) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 abort listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
```


On **Instance 1**: I frequently see the **data looping** in the logs:

```
{"auth":{"applicant_id":"91660679-f0ab-4d1b-b786-f4b3152cfee1","call_retry_number":"1","email":"synch_user@email.com","ip":"::ffff:10.42.1.193","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200782-publicly-reported-operation-interception-demonstrates-ttp-overlaps-with-labyrinth-chollima.pdf","metadata":{"entity_id":"9dae027e-d466-4f8f-abc9-87dd249d2f26","version":"2022-12-08T17:59:18.656Z"},"path":"import/Report/9dae027e-d466-4f8f-abc9-87dd249d2f26"},"timestamp":"2022-12-08T17:59:18.656Z","version":"5.3.17"}
{"auth":{"applicant_id":"91660679-f0ab-4d1b-b786-f4b3152cfee1","call_retry_number":"1","email":"synch_user@email.com","ip":"::ffff:10.42.0.215","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200780-public-report-shows-extent-of-likely-russian-secondary-infektion-disinformation-operations.pdf","metadata":{"entity_id":"7624547a-ee8f-4968-9435-0d8e8f08272f","version":"2022-12-08T18:01:20.907Z"},"path":"import/Report/7624547a-ee8f-4968-9435-0d8e8f08272f"},"timestamp":"2022-12-08T18:01:20.907Z","version":"5.3.17"}
{"auth":{"applicant_id":"91660679-f0ab-4d1b-b786-f4b3152cfee1","call_retry_number":"1","email":"synch_user@email.com","ip":"::ffff:10.42.2.108","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200777-likely-fake-iraqi-paramilitary-groups-make-claims-of-cyber-operation-against-u-s-government-entities-in-iraq.pdf","metadata":{"entity_id":"803d44ee-799c-4752-8544-79176ef92cfa","version":"2022-12-08T18:01:21.701Z"},"path":"import/Report/803d44ee-799c-4752-8544-79176ef92cfa"},"timestamp":"2022-12-08T18:01:21.701Z","version":"5.3.17"}
{"auth":{"applicant_id":"91660679-f0ab-4d1b-b786-f4b3152cfee1","call_retry_number":"1","email":"synch_user@email.com","ip":"::ffff:10.42.1.193","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200779-brazilian-judicial-investigations-ramp-up-with-federal-police-issuing-21-search-and-seizure-warrants.pdf","metadata":{"entity_id":"f48e4dac-a834-4538-ab3b-55db31594a61","version":"2022-12-08T18:01:22.683Z"},"path":"import/Report/f48e4dac-a834-4538-ab3b-55db31594a61"},"timestamp":"2022-12-08T18:01:22.683Z","version":"5.3.17"}
{"auth":{"applicant_id":"91660679-f0ab-4d1b-b786-f4b3152cfee1","call_retry_number":"1","email":"synch_user@email.com","ip":"::ffff:10.42.0.215","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200781-ocean-buffalo-identified-using-highly-targeted-strategic-web-compromise-tactics.pdf","metadata":{"entity_id":"1db6e797-c0b5-4e6a-a8ac-b23f259fb3de","version":"2022-12-08T18:01:24.069Z"},"path":"import/Report/1db6e797-c0b5-4e6a-a8ac-b23f259fb3de"},"timestamp":"2022-12-08T18:01:24.069Z","version":"5.3.17"}
{"auth":{"applicant_id":"91660679-f0ab-4d1b-b786-f4b3152cfee1","call_retry_number":"1","email":"synch_user@email.com","ip":"::ffff:10.42.2.108","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"},"category":"AUDIT","level":"info","message":"UPLOAD","resource":{"filename":"csa-200782-publicly-reported-operation-interception-demonstrates-ttp-overlaps-with-labyrinth-chollima.pdf","metadata":{"entity_id":"9dae027e-d466-4f8f-abc9-87dd249d2f26","version":"2022-12-08T18:01:25.456Z"},"path":"import/Report/9dae027e-d466-4f8f-abc9-87dd249d2f26"},"timestamp":"2022-12-08T18:01:25.456Z","version":"5.3.17"}
```

On **Instance 2** (receiving OpenCTI Instance): I noticed the OpenCTI UI was showing the time slowly increasing and then suddenly showing a value earlier in the day. I checked the logs and found that there would be an error at the same time and then the time state would reset to earlier in the day being processed:

```
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667348029798-0/2022-11-02T00:13:49.798Z","timestamp":"2022-12-08T08:44:25.149Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667348047017-0/2022-11-02T00:14:07.017Z","timestamp":"2022-12-08T08:44:40.278Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667348060379-0/2022-11-02T00:14:20.379Z","timestamp":"2022-12-08T08:44:55.365Z","version":"5.3.17"}
{"category":"APP","error":{"type":"error"},"level":"error","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: error in sync event","timestamp":"2022-12-08T08:46:08.989Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: listening https://opencti.myinstance.com/stream/live?from=2022-11-02T00:14:20.379Z&listen-delete=true&no-dependencies=false&recover=2022-11-23T20:53:14.948Z with id f42a6a47-ab46-4ca8-b0c0-b2ab26e9ab64","timestamp":"2022-12-08T08:46:11.115Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667310650483-0/2022-11-01T13:50:50.483Z","timestamp":"2022-12-08T08:46:11.877Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667310652922-0/2022-11-01T13:50:52.922Z","timestamp":"2022-12-08T08:46:27.448Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667310653682-0/2022-11-01T13:50:53.682Z","timestamp":"2022-12-08T08:46:55.904Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667310689933-0/2022-11-01T13:51:29.933Z","timestamp":"2022-12-08T08:47:10.964Z","version":"5.3.17"}
{"category":"APP","level":"info","message":"[OPENCTI] Sync ad4f4637-a24d-4497-9998-8b26d38d8924: saving state from event to 1667310703271-0/2022-11-01T13:51:43.271Z","timestamp":"2022-12-08T08:47:26.021Z","version":"5.3.17"}
```

**Comment by richard-julien** [comment]
Hi @Marpyre ,

> MaxListenersExceededWarning

Its a elasticsearch client warning, already tracked by 4 or 5 issues on the elastic github.

> Data looping.

Bit difficult to tell you thats happen. As the logs are in Instance 1, it could not come from the synchro, you maybe have connectors that send you a lot of time the same files. To know a bit more about who doing that you should assign a different user for each of your connectors.

> Back in time

For a performance issue we not writting the state every time a message is sent from the stream. So if the stream have any connectivity issue and the connection is reestablish the consumer will maybe reprocess some messages.

**Comment by kmcmahon1959** [comment]
I've seen similar activity with Stream processing (reverting to an earlier time). I suspect this issue is in SSECLIENT, where it will re-"_connect" under certain situations -- using the original URL from OpenCTI vs the last_id processed in the stream. Still testing this issue, as I haven't been able to re-create this since I found this code.

**Comment by Marpyre** [comment]
Hi @richard-julien ,

> Bit difficult to tell you thats happen. As the logs are in Instance 1, it could not come from the synchro, you maybe have connectors that send you a lot of time the same files. To know a bit more about who doing that you should assign a different user for each of your connectors.

I do have a different user for each connector. All of these messages are from the same user, the sync user.

Hi @kmcmahon1959 ,

> I've seen similar activity with Stream processing (reverting to an earlier time). I suspect this issue is in SSECLIENT, where it will re-"_connect" under certain situations -- using the original URL from OpenCTI vs the last_id processed in the stream. Still testing this issue, as I haven't been able to re-create this since I found this code.

It does seem to go back to the original time entered in the synchronizer. Can you link to the code you are referring to?

**Comment by kmcmahon1959** [comment]
In the opencti_connector_helper ListenStream function it calls SSECLIENT (line 394) when it starts up.  In /usr/local/lib/python3.10/site-packages/sseclient.py (line 100) it calls the _connect function when handling certain exceptions -- using the original input url (with the original start time). The sseclient library keeps track of the last_id and would resume from it's last successful read, but the input url overrides that. I have the exception handling code commented out with a "raise" in it's place.  But I haven't had cycles to put into figuring out how to handle those exceptions; OR, better yet, figure out how to let sseclient handle them as it was designed to.

**Comment by richard-julien** [comment]
@Marpyre , can you double check that your worker for the platform that have the synchronizer is NOT talking to the source platform? Its completlty not normal that the sync user that is the user that consume the stream try to upload file on the source platform, it should only interact with the synchronized platform. I will also rework a bit the live stream to have more logs.

@kmcmahon1959 , for the python client reconnection i will take a look.

**Comment by Marpyre** [comment]
@richard-julien 

The source platform (Instance 1) is setup with a Role (named Synch - only permission is Connect and consume the platform streams (/stream, /stream/live),  Group (named Synch - all marking definitions checked), and a User (synch_user@email.com who has the Synch Role and is in the Synch Group).

On the synchronizer instance (Instance 2), the synchronizer is setup with the URL for the source platform (Instance 1) and the token for the synch_user to consume the /live stream from Instance 1.

Please advise if this setup makes sense.

**Comment by Marpyre** [comment]
@richard-julien 

Is it an issue to initialize an OpenCTI instance from an Elasticsearch backup/restore of another OpenCTI instance? To seed the synchronizer instance (Instance 2), we initially copied the Elasticsearch data (dropping the internal OpenCTI indices) to minimize the amount of data being synchronized. Is this valid or will this cause other issues? We tried this approach because we had similar issues when synchronizing all the data and were investigating other ways of moving the data.

**Comment by richard-julien** [comment]
@Marpyre , the setup make sens. What you need to know is that the synchronizer will consume the stream and put every message in an internal rabbitmq queue for the instance2. Then the worker dedicated for the instance 2 will consume this queue to do operations. I dont know your architecture and if you share the rabbitmq, but if its the case you have to use config prefix  to prevent worker to consume data  from an unwanted instance. RabbitMQ prefix config was introduce in the 5.4 if i remember well.

Make an initiailization from a backup to limit amount of sync data to start seems ok for me. Not tested myself but definitly make sens.

**Comment by bigverm23** [comment]
happy to submit any logs here to assist.  I am experiencing the exact same issue with the MaxListeners and then data looping 

2022-12-16T16:35:00.477957261Z INF LOGIN | auth={"email":"xxxx@gmail.com","ip":"::ffff:172.20.0.32","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"} category=AUDIT resource={"provider":"Bearer"} timestamp=2022-12-16T16:35:00.477Z version=5.4.1
2022-12-16T16:35:00.554288513Z INF LOGIN | auth={"email":"xxxxx@gmail.com","ip":"::ffff:172.20.0.27","user_id":"88ec0c6a-13ce-5e39-b486-354fe4a7084f"} category=AUDIT resource={"provider":"Bearer"} timestamp=2022-12-16T16:35:00.554Z version=5.4.1

I am running a single instance of opencti in docker, along with all other containers

**Comment by Marpyre** [comment]
@kmcmahon1959, What you describe seems to be part of the behavior I'm experiencing. When I get a connection error, the synchronizer resets to the original time it started at. 

@richard-julien How do I tell the error that's causing the synchronizer to reset? I just have timestamps of when the error happens in the stream so all I can see is the data that the error happens on. It consistently fails at exactly the same time on every loop.

**Comment by thoyhtya** [comment]
Problem seems to persist in version 5.6.2
Below log sample shows how sync was proceeding normally but after an error it resets to original start time instead of current state that is correctly shown in listening url.

```... "saving state from event to 1656281153301-0/2022-06-26T22:05:53.301Z","timestamp":"2023-03-22T02:31:12.588Z","version":"5.6.2"}
... "saving state from event to 1656281165090-0/2022-06-26T22:06:05.090Z","timestamp":"2023-03-22T02:31:27.646Z","version":"5.6.2"}
... "saving state from event to 1656281179612-0/2022-06-26T22:06:19.612Z","timestamp":"2023-03-22T02:31:42.920Z","version":"5.6.2"}
... "saving state from event to 1656281192543-0/2022-06-26T22:06:32.543Z","timestamp":"2023-03-22T02:31:57.931Z","version":"5.6.2"}
... "error in sync event","timestamp":"2023-03-22T02:32:04.385Z","version":"5.6.2"}
... "listening https://<opencti-url>/stream/<streamid>?from=2022-06-26T22:06:32.543Z&listen-delete=true&no-dependencies=false&recover=2023-03-20T12:09:29.691Z with id 9a996cba-6a9a-46de-a7fe-412abff1bfe9","timestamp":"2023-03-22T02:32:28.582Z","version":"5.6.2"}
... "saving state from event to 1650677161469-0/2022-04-23T01:26:01.469Z","timestamp":"2023-03-22T02:32:32.084Z","version":"5.6.2"}
... "saving state from event to 1650677195764-0/2022-04-23T01:26:35.764Z","timestamp":"2023-03-22T02:32:47.158Z","version":"5.6.2"}
... "saving state from event to 1650677322354-0/2022-04-23T01:28:42.354Z","timestamp":"2023-03-22T02:33:02.563Z","version":"5.6.2"}

**Comment by ristowee** [comment]
This problem is basically the same as in https://github.com/OpenCTI-Platform/client-python/issues/322.
The stream URL is set only once in the EventSource initiation phase. If the connection breaks, the EventSource will reconnect [automatically](https://github.com/EventSource/eventsource/blob/master/lib/eventsource.js#L72). The `connect()` is called again but the url is still the same.
The URL that's shown in the [`[OPENCTI] Sync UUID listening`](https://github.com/OpenCTI-Platform/opencti/blob/5.6.2/opencti-platform/opencti-graphql/src/manager/syncManager.js#L61) print differs from the EventSource internal `url`, which is actually sent to the server. The `eventSource.uri` has no functional meaning.

There are a couple options:
- Remove the `from` field from the URL and set `Last-Event-ID` in the headers at EventSource initiation. If the connection breaks the EventSource sets the `Last-Event-ID` header in the request based on the last received event.
- Modify the EventSource by adding a function that sets the [`reconnectedUrl`](https://github.com/EventSource/eventsource/blob/master/lib/eventsource.js#L62), which will taken into use if reconnetion happens. The place for the function call could be where the [`eventSource.uri`](https://github.com/OpenCTI-Platform/opencti/blob/5.6.2/opencti-platform/opencti-graphql/src/manager/syncManager.js#L111) is set.
